lib/internal/streams/from.js

```diff
@@ -34,21 +34,49 @@ function from(Readable, iterable, opts) {
     objectMode: true,
     ...opts
   });
+
   // Reading boolean to protect against _read
   // being called before last iteration completion.
   let reading = false;
+
+  // needToClose boolean if iterator needs to be explicitly closed
+  let needToClose = false;
+
   readable._read = function() {
     if (!reading) {
       reading = true;
       next();
     }
   };
+
+  readable._destroy = function(error, cb) {
+    close().then(
+      () => cb(error),
+      (e) => cb(error || e),
+    );
+  };
+
+  async function close() {
+    if (needToClose) {
+      needToClose = false;
+      if (typeof iterator.return === 'function') {
+        const { value } = await iterator.return();
+        await value;
+      }
+    }
+  }
+
```

Comment:
This should not be an async function. Mixing callbacks and async creates weird/unexpected error scenarios.

Reply:
> creates weird/unexpected error scenarios

It does? In what way?

Reply:
@vadzim sorry, but I'm not sure this comment is resolved, @mcollina?

Reply:
@ronag I've rebased and squashed commits as mentioned in PR requirements.
In one of previous version `_destroy` was async